components/style/values/computed/image.rs

```diff
@@ -0,0 +1,291 @@
+        });
+        Ok(())
+    }
+}
+
+/// Computed values for EndingShape
+#[derive(Clone, PartialEq)]
+#[cfg_attr(feature = "servo", derive(HeapSizeOf))]
+pub enum EndingShape {
+    Circle(Size<Length>),
+    Ellipse(Size<Length>, Size<Length>),
+}
+
+impl ::cssparser::ToCss for EndingShape {
+    fn to_css<W>(&self, dest: &mut W) -> fmt::Result where W: fmt::Write {
+        match *self {
+            EndingShape::Circle(ref length) => {
+                try!(dest.write_str("circle "));
+                try!(length.to_css(dest));
+            },
+            EndingShape::Ellipse(ref first_len, ref second_len) => {
+                try!(dest.write_str("ellipse "));
+                try!(first_len.to_css(dest));
+                try!(dest.write_str(" "));
+                try!(second_len.to_css(dest));
+            },
+        }
+        Ok(())
+    }
+}
+
+impl fmt::Debug for EndingShape {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            EndingShape::Circle(ref length) => {
+                let _ = write!(f, "circle {:?}", length);
+            },
+            EndingShape::Ellipse(ref first_len, ref second_len) => {
+                let _ = write!(f, "ellipse {:?} {:?}", first_len, second_len);
+            }
+        }
+        Ok(())
+    }
+}
+
+#[derive(Clone, PartialEq)]
+#[cfg_attr(feature = "servo", derive(HeapSizeOf))]
+pub enum Size<T: Debug + LocalToCss> {
+    Length(T),
+    Keyword(SizeKeyword),
+}
+
+impl<T: Debug + LocalToCss> ::cssparser::ToCss for Size<T> {
+    fn to_css<W>(&self, dest: &mut W) -> fmt::Result where W: fmt::Write {
+        match *self {
+            Size::Length(ref length) => length.to_css(dest),
+            Size::Keyword(keyword) => keyword.to_css(dest),
+        }
+    }
+}
+
+impl<T: Debug + LocalToCss> fmt::Debug for Size<T> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            Size::Length(ref length) => {
+                let _ = write!(f, "{:?}", length);
+            },
+            Size::Keyword(keyword) => {
+                let _ = write!(f, "{:?}", keyword);
+            },
+        }
+        Ok(())
+    }
+}
+
+impl ToComputedValue for specified::image::Gradient {
+    type ComputedValue = Gradient;
+
+    #[inline]
+    fn to_computed_value(&self, context: &Context) -> Gradient {
+        let specified::image::Gradient {
+            ref stops,
+            repeating,
+            ref gradient_kind
+        } = *self;
+        Gradient {
+            stops: stops.iter().map(|stop| {
+                ColorStop {
+                    color: stop.color.parsed,
+                    position: match stop.position {
+                        None => None,
+                        Some(value) => Some(value.to_computed_value(context)),
+                    },
+                }
+            }).collect(),
+            repeating: repeating,
+            gradient_kind: gradient_kind.clone()
```

Comment:
This will become `gradient_kind.to_computed_value(context)` once you create computed gradient values.

It's fine to split these ToComputedValue impls out, btw, no need to have it all in one function. I know that's how it was before, but that's not a guideline we need to follow, necessarily.

Use `ComputedValueAsSpecified` on the keyword enums, btw.